pkg/api/authn_test.go

```diff
@@ -866,6 +867,327 @@ func TestAPIKeys(t *testing.T) {
+		selfSignedClientKeyPath := path.Join(tempDir, "client-selfsigned.key")
+		err = tlsutils.GenerateSelfSignedCertWithCNToFile("testuser", selfSignedClientCertPath, selfSignedClientKeyPath)
+		So(err, ShouldBeNil)
+
+		// Set up server
+		conf := config.New()
+		port := test.GetFreePort()
+		baseURL := test.GetSecureBaseURL(port)
+
+		conf.HTTP.Port = port
+		conf.HTTP.TLS = &config.TLSConfig{
+			Cert:   serverCertPath,
+			Key:    serverKeyPath,
+			CACert: caCertPath,
+		}
+		conf.HTTP.AccessControl = &config.AccessControlConfig{
+			Groups: config.Groups{
+				"mtls-users": config.Group{
+					Users: []string{"testuser"},
+				},
+			},
+			Repositories: config.Repositories{
+				"**": config.PolicyGroup{ // Default restrict all
+					AnonymousPolicy: make([]string, 0),
+					Policies:        make([]config.Policy, 0),
+				},
+				"test-repo": config.PolicyGroup{
+					Policies: []config.Policy{
+						{
+							Users:   []string{"testuser"},
+							Actions: []string{"read", "create"},
+						},
+					},
+				},
+			},
+		}
+		conf.Storage.RootDirectory = t.TempDir()
+
+		ctlr := api.NewController(conf)
+		cm := test.NewControllerManager(ctlr)
+
+		cm.StartAndWait(port)
+		defer cm.StopServer()
+
+		Convey("mTLS only authentication", func() {
+			// Test without client certificate - should fail
+			caCertPEM, err := os.ReadFile(caCertPath)
+			So(err, ShouldBeNil)
+
+			caCertPool := x509.NewCertPool()
+			caCertPool.AppendCertsFromPEM(caCertPEM)
+
+			client := resty.New()
+			client.SetTLSClientConfig(&tls.Config{RootCAs: caCertPool, MinVersion: tls.VersionTLS13})
+			_, err = client.R().Get(baseURL + "/v2/test-repo/tags/list")
+			So(err, ShouldNotBeNil)
+			So(err.Error(), ShouldContainSubstring, "tls: certificate required")
+
+			// Test with valid client certificate - should succeed
+			clientCert, err := tls.LoadX509KeyPair(clientCertPath, clientKeyPath)
+			So(err, ShouldBeNil)
+
+			client = resty.New()
+			client.SetTLSClientConfig(&tls.Config{
+				MinVersion:   tls.VersionTLS13,
+				Certificates: []tls.Certificate{clientCert},
+				RootCAs:      caCertPool,
+			})
+
+			resp, err := client.R().Get(baseURL + "/v2/test-repo/tags/list")
+			So(err, ShouldBeNil)
+			So(resp.StatusCode(), ShouldEqual, http.StatusNotFound) // 404 meaning we successfully passed auth
+
+			// Test with self-signed client certificate - should fail
+			selfSignedClientCert, err := tls.LoadX509KeyPair(selfSignedClientCertPath, selfSignedClientKeyPath)
+			So(err, ShouldBeNil)
+
+			client = resty.New()
+			client.SetTLSClientConfig(&tls.Config{
+				MinVersion:   tls.VersionTLS13,
+				Certificates: []tls.Certificate{selfSignedClientCert},
+				RootCAs:      caCertPool,
+			})
+
+			_, err = client.R().Get(baseURL + "/v2/test-selfsigned-repo/tags/list")
+			So(err, ShouldNotBeNil)
+			So(err.Error(), ShouldContainSubstring, "tls: certificate required")
+		})
+	})
+
+	Convey("Test mTLS with basic auth authentication", t, func() {
```

Comment:
According to my understanding of the Convey execution order (https://github.com/smartystreets/goconvey/wiki/Execution-order), the 3 nested Conveys under this one would re-generate the certificates for each test. I'm wondering if the certificates could be generated outside the Convey and then re-used for the tests.